Configure backup path in prefs

Siedlerchr opened the pull request:
Adds a new option in the preferences to enable/disable backups and to choose the backup directory
Fallback/Default dir is always app dir as usual

- [x] Enable/disable backup
- [x] Fill the textfield in the prefs 


<!-- 
Describe the changes you have made here: what, why, ... 
Link issues that are fixed, e.g. "Fixes #333".
If you fixed a koppor issue, link it, e.g. "Fixes https://github.com/koppor/jabref/issues/47".
The title of the PR must not reference an issue, because GitHub does not support autolinking there.
-->


<!-- 
- Go through the list below. Please don't remove any items.
- [x] done; [ ] not done / not applicable
-->

```[tasklist]
### Compulsory checks
- [ ] Change in `CHANGELOG.md` described in a way that is understandable for the average user (if applicable)
- [ ] Tests created for changes (if applicable)
- [ ] Manually tested changed features in running JabRef (always required)
- [ ] Screenshots added in PR description (for UI changes)
- [ ] [Checked developer's documentation](https://devdocs.jabref.org/): Is the information available and up to date? If not, I outlined it in this pull request.
- [ ] [Checked documentation](https://docs.jabref.org/): Is the information available and up to date? If not, I created an issue at <https://github.com/JabRef/user-documentation/issues> or, even better, I submitted a pull request to the documentation repository.
```

koppor commented on src/main/java/org/jabref/logic/autosaveandbackup/BackupManager.java:
```diff
@@ -59,18 +59,19 @@ public class BackupManager {
     private final Queue<Path> backupFilesQueue = new LinkedBlockingQueue<>();
 
-    private boolean needsBackup = false;
+    private boolean needsBackup = true;
```
Why is this boolean changed? This would revert https://github.com/JabRef/jabref/pull/9679 and thus probably reopen https://github.com/JabRef/jabref/issues/9488#issuecomment-1445149783.

Siedlerchr replied:
Oh good catch! Not intended

koppor replied:
Maybe, we should add a link to the PR and the issue there? - I also put `true` at the beginning to have the backup dir filled during development... -- I think, a short link at the comment is better than to say: "Look at git blame and read on there to understand this line of code ^^"

koppor commented on src/main/java/org/jabref/logic/util/io/BackupFileUtil.java:
```diff
@@ -48,12 +48,13 @@ public static Path getAppDataBackupDir() {
         String extension = "." + fileType.getExtensions().get(0);
         String timeSuffix = ZonedDateTime.now().format(DateTimeFormatter.ofPattern("yyyy-MM-dd--HH.mm.ss"));
 
         // We choose the data directory, because a ".bak" file should survive cache cleanups
```
This comment can be removed here (I left a comment, where it belongs now)

koppor commented on src/main/java/org/jabref/logic/util/io/BackupFileUtil.java:
```diff
@@ -48,12 +48,13 @@ public static Path getAppDataBackupDir() {
         String timeSuffix = ZonedDateTime.now().format(DateTimeFormatter.ofPattern("yyyy-MM-dd--HH.mm.ss"));
 
         // We choose the data directory, because a ".bak" file should survive cache cleanups
-        Path directory = getAppDataBackupDir();
+        Path directory = backupDir;
```
Can you rename the local variable `directory` to `backupDir` and remove this statement?

koppor commented on src/main/java/org/jabref/logic/util/io/BackupFileUtil.java:
```diff
@@ -66,42 +67,37 @@ public static Path getPathForNewBackupFileAndCreateDirectory(Path targetFile, Ba
-    public static Optional<Path> getPathOfLatestExistingBackupFile(Path targetFile, BackupFileType fileType) {
-        // The code is similar to "getPathForNewBackupFileAndCreateDirectory"
+    public static Optional<Path> getPathOfLatestExistingBackupFile(Path targetFile, BackupFileType fileType, Path backupDir) {
+            // The code is similar to "getPathForNewBackupFileAndCreateDirectory"
```
Indent seems to be off here?

koppor commented on src/main/java/org/jabref/preferences/JabRefPreferences.java:
```diff
@@ -2172,8 +2179,9 @@ public FilePreferences getFilePreferences() {
-        );
+                ExternalFileTypes.fromString(get(EXTERNAL_FILE_TYPES)),
+                getBoolean(CREATE_BACKUP),
+                getPath(BACKUP_DIRECTORY, BackupFileUtil.getAppDataBackupDir()));
```
Please add the comment from above to here:

```java
// We choose the data directory, because a ".bak" file should survive cache cleanups
```

koppor reviewed (changes requested):
Some minor comments.

Yeah, the enable and disable of the backup (start/stop of BackupManager with other configuraiton) really needs to be done.

HoussemNasri commented:
I checkout the PR, but then I can't seem to trigger the backup manager dialog. I have tried closing JabRef from IntelliJ and from the Taskmanager by ending the process but neither has worked; when I start JabRef again, the backup dialog doesn't show. Keep in mind that the "Create backup" is enabled.

polish: besides that problem, I think the backup path text field should be disabled when the "Create backup" checkbox is unchecked.

koppor commented:
> I checkout the PR, but then I can't seem to trigger the backup manager dialog. I have tried closing JabRef from IntelliJ and from the Taskmanager by ending the process but neither has worked;

Did you also modify the library without saving?

I did and I got the dialog:

![image](https://user-images.githubusercontent.com/1366654/236352093-d7b01d2e-b3c0-4292-b92a-875ab71cdcde.png)

> polish: besides that problem, I think the backup path text field should be disabled when the "Create backup" checkbox is unchecked.

Current state:

![image](https://user-images.githubusercontent.com/1366654/236352186-07e57cdf-8871-4773-ba07-4b1521a8feb4.png)

We should also have a button "reset to default" (same icon as in "Reset preferences")

HoussemNasri commented:
Yes, I modified the library before closing it. I also changed the backup path.

Siedlerchr commented:
A restart might be necessary after changing the prefs, the back up manager
is started very early

Houssem Nasri ***@***.***> schrieb am Fr., 5. Mai 2023, 01:51:

> Yes, I modified the library before closing it. I also changed the backup
> path.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/JabRef/jabref/pull/9809#issuecomment-1535536398>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AACOFZEE6UOQD53GWPFAJTTXEQ6IRANCNFSM6AAAAAAXOGJY5M>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

HoussemNasri commented:
@Siedlerchr It is working now! A restart was needed as you said. Also, I noticed that if I open JabRef, make a change, and then close it *(kill the process)* very fast, the backup dialog doesn't show the next time.

Siedlerchr commented:
Yep, you have to wait a few seconds because the backup path is using seconds in the filename only as minimum        `String timeSuffix = ZonedDateTime.now().format(DateTimeFormatter.ofPattern("yyyy-MM-dd--HH.mm.ss"));`
 
 That's why I added a sleep in the test as well

HoussemNasri commented on src/main/java/org/jabref/gui/JabRefFrame.java:
```diff
@@ -490,7 +490,7 @@ public boolean quit() {
             }
             AutosaveManager.shutdown(context);
-            BackupManager.shutdown(context);
+            BackupManager.shutdown(context, prefs.getFilePreferences().getBackupDirectory(), prefs.getFilePreferences().shouldCreateBackup());
```
thought: three arguments is a bit too much for a static method. I think it is time to convert BackupManager to use instance methods.

Siedlerchr replied:
The backup manager is started early for each bib file. That cannot be changed on the fly. Maybe add this to the restart warnings 

The BackupManager is a combination of static and instance methods. Basically, the BackupManager has a static set with all running instances.  Start adds a new instance of the backup manager. A backup manager is created for each bibdatabasecontext and the running instances set ensures that there is at most one for each bib file.

HoussemNasri commented on src/main/java/org/jabref/logic/util/io/BackupFileUtil.java:
```diff
@@ -66,18 +67,12 @@ public static Path getPathForNewBackupFileAndCreateDirectory(Path targetFile, Ba
 
         String extension = "." + fileType.getExtensions().get(0);
 
-        Path directory = getAppDataBackupDir();
```
nitpick: getAppDataBackupDir could be renamed to getDefaultBackupDir

Siedlerchr replied:
That method is still valid and does return the AppDataBackupDir. However, I found another method where I did not pass the correect param

HoussemNasri reviewed (changes requested):
LGTM codewise, I didn't look too much into the tests but they look good overall.  It would be great if you can get the backup directory to be updated without restarting, otherwise, a "you need to restart" message is fine.
